Nix language tutorial (#267)
NixOS/nix.dev: source/tutorials/nix-language.md

Example:

```nix
{a, b}: a + b
```

    <LAMBDA>

The argument defines the exact attributes that have to be in that set.
Leaving out or passing additional attributes is an error.

Example:

```nix
let
  f = {a, b}: a + b;
in
f { a = 1; b = 2; }
```

    3

Counter-example:

```nix
let
  f = {a, b}: a + b;
in
f { a = 1; b = 2; c = 3; }
```

    error: 'f' at (string):2:7 called with unexpected argument 'c'

           at «string»:4:1:

                3| in
                4| f { a = 1; b = 2; c = 3; }
                 | ^
                5|

<!-- TODO: not the same as x: x.a + x.b (!!!!) -->

### Default attributes

Also known as “default arguments”.

Destructured arguments can have default values for attributes.

This is denoted by separating the attribute name and its default value with a question mark (`?`).

Attributes in the argument are not required if they have a default value.

Example:

```nix
let
  f = {a, b ? 0}: a + b
in
f { a = 1; }
```

    1

Example:

```nix
let
  f = {a ? 0, b ? 0}: a + b
in
f { } # empty attribute set
```

    0

### Additional attributes

> Additional attributes are allowed with an ellipsis (`...`):

Reviewer: Just read through this and this section was the only thing to give me pause. I already knew this but I think bringing a little more attention to the use of ellipsis would be good.

```suggestion
Additional attributes are allowed with an ellipsis (`...`).

Unlike the previous counter-example, passing additional arguments will not be an error.

Example:
```

    {a, b, ...}: a + b

> Example:

Reviewer: Just read through this and this section was the only thing to give me pause. I already knew this but I think bringing a little more attention to the use of ellipsis would be good.

```suggestion
Unlike the previous counter-example, passing additional arguments will not be an error.

Example:
```

(reading-nix-language)=

# Nix language basics

The Nix language is used to declare packages and configurations for [Nix][nix-manual].

It is a domain-specific, purely functional, lazily evaluated, dynamically typed programming language.

Notable uses of the Nix language are:

- [Nixpkgs][nixpkgs]

>   The largest, most up-to-date software distribution in the world, and written in Nix language.

Author: I tried to leave out articles where they appeared not strictly necessary. This is up for debate of course, but if we can have fewer words that would be great.

>   The largest, most up-to-date software distribution in the world, and written in Nix language.

Reviewer 2: I agree with @matta that not having a `the` here feels very unnatural

- [NixOS][nixos-manual]

  Linux distribution based on Nix, which can be configured fully declaratively.

  Its underlying modular configuration system is written in Nix language, and uses packages from Nixpkgs.
  The operating system environment and services it provides are configured with Nix language.

[nix-manual]: https://nixos.org/manual/nix/stable
[nixpkgs-manual]: https://nixos.org/manual/nixpkgs/stable/#preface
[nixos-manual]: https://nixos.org/manual/nixos/stable/index.html#preface
[home-manager]: https://github.com/nix-community/home-manager

## Overview

Using the Nix language in practice entails multiple things:

- language: syntax and semantics
- function libraries: `builtins` and `pkgs.lib`
- developer tools: testing, debugging, linting and formatting, ...
- generic build mechanisms: `stdenv`, trivial builders, ...
- composition and configuration mechanisms: `override`, `overrideAttrs`, overlays, `callPackage`, ...

> - ecosystem-specific packaging mechanisms: `buildGoModule`, `buildPythonApplication`, ...

Author: Yes. Why would one capitalize this? They are not sentences, not even close.

> - ecosystem-specific packaging mechanisms: `buildGoModule`, `buildPythonApplication`, ...

Reviewer 2: Ah yeah makes sense

- NixOS module system: `config`, `option`, ...

**This guide only covers language syntax and semantics**, briefly discusses function libraries, and at the end will direct you to resources on the other components.

### What will you learn?

This guide should enable you to read typical Nix language code and understand its structure.

It shows the most common and distingushing patterns in the Nix language:

- [assigning names and accessing values](names-values)
- declaring and calling [functions](functions)
- [built-in and library functions](libraries)
- [side effects](side-effects) to obtain build inputs and produce build results

It *does not* explain all Nix language features in detail.
See the [Nix manual][manual-language] for a full language reference.

[manual-language]: https://nixos.org/manual/nix/stable/language/index.html

### What do you need?

- Familiarity with software development
- Familiarity with Unix shell, to read command line examples <!-- TODO: link to yet-to-be instructions on "how to read command line examples" -->

> - [](./install-nix), to run the examples

Reviewer 3: There is a missing text here

Author: Grammatically, you are probably right.

### How long does it take?

- **No experience with functional programming: 2 hours**

  Carefully work through all definitions and examples.

  Make sure you understand them by reading all detailed explanations.

- **Familiar with functional programming: 1 hour**

  Learn syntax and semantics by examining definitions and examples, and read detailed explanations if in doubt.

- **Proficient with functional programming: 30 minutes**

  The Nix language is largely unsurprizing to experienced functional programmers.

  Read through the guide to get accustomed to the syntax and language features.

We recommend running all examples and playing with them to validate your assumptions and test what you have learned.

### How to run the examples?

- Every valid piece of Nix language code is a *Nix expression*.

> - Nix expressions can contain other Nix expressions, that is, they can be nested.

Author: That was indeed a fix after running it against beginners, clarifying "Nix file = Nix expression". I originally took that note from @tazjin's nix-1p, because I would have found that information very valuable early on. I can clearly remember wondering what kind of thing was in a Nix file, and someone telling me that it evaluates to exactly one thing would have helped a lot.

But, as you say, it made people wonder if expressions can be nested, and indeed they can, so that appeared like a reasonable addition.

So while I agree that it may be too much information at this point, I'm inclined to listen beginners here rather than experts.

> - Nix expressions can contain other Nix expressions, that is, they can be nested.

Reviewer 2: I'm not talking about "Evaluating a Nix expression produces a single value.", that makes a lot of sense and I can see how that helps beginners. But the fact that expressions can be nested feels out of place here. If we need to point out more explicitly that expressions can contain other expressions, this might be better further down somewhere.

> - Evaluating a Nix expression produces a single value.

Author: Good catch!

- Every *Nix file* (`.nix`) contains a single Nix expression.

:::{note}
To *evaluate* means to transform an expression according to the language rules until no further simplification is possible.
:::

All examples in this guide are valid Nix expressions.

> Each one is followed by the expected evaluation result.

Author: Right, that seems to be an artifact before counter-examples came into play. Do you have a suggestion to fix this?

> Each one is followed by the expected evaluation result.

Reviewer 2: How about "This guide contains examples of Nix expressions, each followed by the expected evaluation result"

The following example is a Nix expression adding two numbers:

```nix
1 + 2
```

    3

#### Interactive evaluation

> Use [`nix repl`][nix-repl] to evaluate Nix expressions interactively (by typing them on the command line):

Reviewer 2: You need to mention `:p` for strict evaluation somewhere, it's needed for many examples

Author: Not strictly needed, but yes, sometimes otherwise the output will not be the same.

```console
nix repl
```

    Welcome to Nix 2.5.1. Type :? for help.

```console
nix-repl> 1 + 2

> ```

Author: This will be distiguished visually with https://github.com/NixOS/nix.dev/pull/320

    3

[nix-repl]: https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-repl.html

#### Evaluating Nix files

Use [`nix-instantiate --eval`][nix-instantiate] to evaluate the expression in a Nix file.

```console
echo 1 + 2 > file.nix

nix-instantiate --eval file.nix
```

    3

<details><summary>Detailed explanation</summary>

The first command writes `1 + 2` to a file `file.nix` in the current directory.
The contents of `file.nix` are now `1 + 2`, which you can check with

```console
cat file.nix
```

    1 + 2

The second command runs `nix-instantiate` with the `--eval` option on `file.nix`, which reads the file and evaluates the contained Nix expression.
The resulting value is printed as output.

`--eval` is required to evaluate the file and do nothing else.

> If `--eval` is omitted, `nix-instantiate` expects the expression in the given file to evaluate to a special data type called *Derivation*, which we will come back to at the end of this guide in [](build-tasks).

Author: I'd argue it is special, because it's the whole point of Nix. It deserves special attention. Even if we should decide not to call it special here, I would like to emphasize that it is a value of a certain data type as opposed to the other three things a {store d,d,`d`}erivation could be.

> If `--eval` is omitted, `nix-instantiate` expects the expression in the given file to evaluate to a special data type called *Derivation*, which we will come back to at the end of this guide in [](build-tasks).

Reviewer 2: Fair enough, I mainly think "data type" is not right. How about

```suggestion
If `--eval` is omitted, `nix-instantiate` expects the expression in the given file to evaluate to a special value called *Derivation*, which we will come back to at the end of this guide in [](build-tasks).
```

If you do not need `file.nix` any more, remove it with

```console
rm file.nix
```

</details>

:::{note}
`nix-instantiate --eval` will evaluate `default.nix` if no file name is specified.

```console
echo 1 + 2 > default.nix

nix-instantiate --eval
```

    3
:::

[nix-instantiate]: https://nixos.org/manual/nix/stable/command-ref/nix-instantiate.html

## Reading the Nix language without fear

> You will quickly encounter Nix language expressions that may look very complicated.

Author: I didn't want to judge if they *actually* appear complicated to the reader, but I'm pretty sure one will encounter such things.

> You will quickly encounter Nix language expressions that may look very complicated.

Reviewer 2: The _may_ at the beginning should already take care of this, the user may encounter expressions that look complicated [to them]. "looking complicated" is already subjective. Anyways not a big concern, the original sentence is just a bit off imo

As with any programming language, the required amount of Nix language code closely matches the complexity of the problem it is supposed to solve, and reflects how well the problem – and its solution – is understood.

Building software is a complex undertaking, and Nix both exposes and allows managing this complexity with the Nix language.

> The purpose of the Nix language is to create *build tasks*: precise descriptions of how contents of existing files are used to derive new files.

Author: This would be missing the fact that Nix makes that complexity visible in the first place. (Thanks to @AlexeiDrake for pointing this out, the two paragraphs are almost verbatim his.)

It would also be missing the purpose of the language, which motivates why one would even bother using or having it in the first place, and also builds an arch (of dramatic suspense 🙄) to the end of the tutorial, where we finally get to those build tasks we got promised.

:::{important}
A build task in Nix is called *derivation*.
:::

The Nix language has only few basic constructs which can be combined arbitrarily:

- primitive data types

>   as basic building blocks

Reviewer 2: I guess by this you mean "Expressions representing their values directly", so e.g. `null`, `1`, `"str"`, `{ x = 10; }` but not `1 + 1`, `let x = 10; in x`, etc.

However, this category then also includes functions, e.g. `x: x` or `{ f = x: x; }`, which also evaluate to their values directly. But functions are in the next category already

Author: Yes, functions are also primitives, but as with the other things you pointed out this is purely for helping orient the reader by providing consistent structure in both overview and sections.

Author: > I guess by this you mean "Expressions representing their values directly"

Actually no, I'm not even thinking much in terms of expressions and values in this context, but instead want to address what I suppose are semantics: what do you *do* with these language objects?

In that sense (which is borrowed from Abelson & Sussman's SICP) functions are not primitives, even if the expressions declaring them do not evaluate further. They are means of combination. Bindings in function parameters are a bit more tricky, but intuitively it's not strictly a means of abstraction.

Maybe I'm relying on too ancient sources here, please enlighten me.

Reviewer 2: I think this might be important from what you linked:

> In programming, we deal with two kinds of elements: functions and data. (Later we will discover that they are really not so distinct.)

Because in functional programming languages, functions are first-class citizens and can be passed around like any other data. This then makes the distinction between data and functions tricky, because functions can be both. In the end I think it's really just a matter of convention about how terms are defined, but this leads to a fuzzy categorization. But I get your idea of making this distinction for having a better tutorial flow.

Taking that into account though, I think it would be good to categorize it into "simple expressions" and "function declaration and application"

- compound data types and functions

>   to produce and transform complex data

Reviewer 2: I'd say this category should be everything else, maybe call it "Expressions that aren't fully evaluated", including things like `(x: x) 1`, `let x = 10; in x`, etc. Evaluating all of these gives a value that's different from the expression

Author: I understand what you mean but I really doubt it would help the cause here. For me that kind of wording crosses the boundary of being so technically correct as to confuse unsuspecting beginners.

Sure, being not technically correct will confuse unsuspecting PLT experts. But you're not the intended audience. I'm already taking some liberty with that entire narrative, comparing Nix to JSON, building complex data and all that. These things are just to support that narrative, to create a mental model that is deliberately simplified to focus on the essentials.

I'd love to make it technically correct and still focused and readable, but I'm out of fantasy right now and would love to see a full suggestion how to get there. Feel free to make a PR against this PR, or propose such a significant change after merging.

- name assignment

>   to manipulate complex data as units

Reviewer 2: And then this category falls away. Notably with this category there would be another ambiguity, because functions also introduce names (function arguments).

In addition it allows three side effects:

  - reading files as Nix expressions

>     to enable code reuse

Author: I agree and I was not sure whether to include this, but it motivates to show `import` and gives it a natural place to live, which was otherwise hard to find.

>     to enable code reuse

Reviewer 2: Side effects are kind of a bad thing in purely functional languages, so you'd want to avoid them whenever possible. So I really don't think something that's not even an actual side effect should be advertised as one. There has to be a better place to mention `import`

>     to enable code reuse

Author: Yeah you're right. I'll have to take some time to think about restructuring according to the larger points in the review comments.

  - reading files as *build inputs*

>     to capture what build tasks will operate on

Author: Sure it reads files, otherwise Nix couldn't compute a store path.

>     to capture what build tasks will operate on

Reviewer 2: Ahh you're right, my mistake. (though I don't think it's a relevant side effect)

>     to capture what build tasks will operate on

Author: It is a relevant side effect because files may change between evaluations as opposed to say fetched resources that have a predeclared hash.

  - writing files as *build tasks*

>     to keep them for later execution, the *build*

Author: This is all true, but then again how do we otherwise structure this introduction and still show the relevant mechanisms?

Given the [purpose of the language](https://github.com/notifications?query=backport#discussion_r967533619) it would be pointless to evaluate it without capturing the store derivations. Nix language is not a general-purpose language, there is no point using it outside of that problem domain.

>     to keep them for later execution, the *build*

Reviewer 2: (The link you gave doesn't work)

Derivations should be described as the _output_ of a Nix evaluation (it's a somewhat special type of Nix value), not a side effect. So Nix evaluation happens, a value gets produced, which is then used by other parts of Nix to turn them into something `nix-instantiate` (without `--eval`) creates the derivation in the store, and coupling that with `nix-build` also builds it. But the Nix language is not involved past evaluation.

In such cases, the use of names from the surrounding Nix expression is a coincidence.

Example:

```nix
let
  out = "Nix"
in
"echo ${out} > $out"
```

    "echo Nix > $out"
:::

<!-- TODO: link to escaping rules -->

(functions)=
## Functions

Functions are everywhere in the Nix language and deserve particular attention.

### Single argument

Functions in Nix language can appear in different forms, but always take exactly one argument.
Argument and function body are separated by a colon (`:`).

Wherever you see a colon (`:`) in Nix language code:
- on its left is the function argument
- on its right is the function body.

Nix functions have no names.
We say they are anonymous, and call such a function a *lambda*.

Example:

```nix
x: x + 1
```

    <LAMBDA>

The `<LAMBDA>` indicates the resulting value is an anonymous function.

We can assign functions a name as to any other value.

Example:

```nix
let
  f = x: x + 1
in f
```

    <LAMBDA>

### Calling functions

Also known as "function application".

Calling a function with an operand means writing the operand after the function.

Example:

```nix
(x: x + 1) 1
```

    2

:::{note}

> Since function and operand are separated by white space, sometimes parantheses (`( )`) are required to distinguish expressions.

Author: Since the goal of the tutorial is to read the language, not to write it, I'm convinced it doesn't matter much where semicolons go. But I see what you mean, and will think about how to address these issues.

To answer your immediate question: Semicolons are used for three things.
- delimit attributes in a attribute sets
- delimit `let` declarations
- separate source and sink in `with`

:::

Example:

```nix
let
  f = x: x.a;
in
f { a = 1; }
```

    1

The above example calls `f` on a literal attribute set.
One can also pass operands by name.

Example:

```nix
let
  f = x: x.a;
  v = { a = 1; };
in
f v
```

    1

### Chaining arguments

Arguments can be chained by nesting functions.

Such a nested function can be used like a function that takes multiple arguments, but offers additional flexibility.

Example:

```nix
x: y: x + y
```

    <LAMBDA>

> The above function takes one argument and returns a function `y: x + y` with `x` set to the value of the argument.

Reviewer 4: I am unfamiliar with the phrase "chaining arguments". The first thing I do when faced with this is try to research further, but I found it hard to find any references for what is meant by this phrase, whether it is a phrase I should be using in communication with other people, etc. I did web searches to try to figure out if this phrase is specific to Nix, functional programming, etc., and came up short. I did find results that refer to chaining *functions* but not arguments. This suggests to me that perhaps a tutorial should not use this phrase (unless it is used elsewhere in Nix reference material).

If the goal is to introduce Nix using familiar and clear concepts, perhaps title this section "Handling multiple arguments" and then explain in the section content that Nix has no such thing but a similar thing can be accomplished by (insert common functional programming term here).

In particular I don't quite know how to understand `x: y: x + y` at first glance. Beginning by explaining that the expression is equivalent to `x: (y: y + x)` would have helped me. It is obvious in retrospect, but I am still learning! :-) For some reason the parentheses and placing the `x` last also helps me understand. Or, phrasing the sentence "The above function takes one argument `x` and returns a function `y: y + x` with `x` set to the value of the argument.

Author: > chaining arguments

Yes, you're right. I kind of made that up. I thought about calling it currying, on the one hand this is already quite technical jargon, and on the other hand it's not even right, as currying is defined as transforming higher-arity functions to nested unary functions.

> "Handling multiple arguments" and then explain in the section content that Nix has no such thing

That's what I'll do.

> Beginning by explaining that the expression is equivalent to x: (y: y + x) would have helped me.

I'll definitely keep arguments in the same order, but I'll add the parenthesized version, it's a good idea.

    sha256 = "1ayhp9v4m4rdhjmnl2bq3cibrbqqkgjbl3s7yk2nhlh8vj3ay16g";
  };

  meta = with lib; {
    license = licenses.gpl3Plus;
  };

}
```

This example is a (simplified) package declaration from Nixpkgs.

Explanation:

- This expression is a function that takes an attribute set which must have exactly the attributes `lib` and `stdenv`.
- It returns the result of evaluating the function `mkDerivaion`, which is an attribute of `stdenv`, applied to a recursive set.
- The recursive set passed to `mkDerivation` uses its own `pname` and `version` attributes in the argument to the built-in function `fetchTarball`.
- The `meta` attribute is itself an attribute set, where the `license` attribute has the value that was assigned to the nested attribute `lib.licenses.gpl3Plus`.

## References

- [Nix manual: Nix language][manual-language] - Nix language reference
- [Nix manual: Built-in Functions][nix-builtins] - Nix language built-in functions
- [Nixpkgs manual: Functions reference][nixpkgs-functions] - Nixpkgs function library
- [Nixpkgs manual: Fetchers][nixpkgs-fetchers] - Nixpkgs fetcher library

## Next steps

To get things done:

> - [](declarative-reproducible-envs) – create reproducible shell environments from a Nix file

Reviewer 4: FYI `[](declarative-reproducible-envs)` renders as the empty string in the document.

- Language: syntax and semantics
- Libraries: `builtins` and `pkgs.lib`
- Developer tools: testing, debugging, linting, formatting, ...
- Generic build mechanisms: `stdenv.mkDerivation`, trivial builders, ...
- Composition and configuration mechanisms: `override`, `overrideAttrs`, overlays, `callPackage`, ...
- Ecosystem-specific packaging mechanisms: `buildGoModule`, `buildPythonApplication`, ...
- NixOS module system: `config`, `option`, ...

**This guide only covers some language syntax and mostly semantics**, briefly discusses libraries, and at the end will direct you to resources on the other components.

### What will you learn?

This guide should enable you to read typical Nix language code and understand its structure.

It shows the most common and distingushing patterns in the Nix language:

- [Assigning names and accessing values](names-values)
- Declaring and calling [functions](functions)
- [Built-in and library functions](libraries)
- [Side effects](side-effects) to obtain build inputs
- [Derivations](derivations) that describe build tasks

:::{important}
This guide *does not* explain all Nix language features in detail.

> See the [Nix manual][manual-language] for a full language reference.

Author: I generously and deliberately apply wishful thinking w.r.t. the manual. We (as a community) will fix forward once someone gets around to it.

Read detailed explanations if you want to make sure you fully understand the examples.

### What do you need?

- Familiarity with software development
- Familiarity with Unix shell, to read command line examples <!-- TODO: link to yet-to-be instructions on "how to read command line examples" -->
- A [Nix installation](./install-nix) to run the examples

### How to run the examples?

- A piece of Nix language code is a *Nix expression*.
- Evaluating a Nix expression produces a *Nix value*.
- The content of a *Nix file* (file extension `.nix`) is a Nix expression.

:::{note}
To *evaluate* means to transform an expression into a value according to the language rules.
:::

This guide contains many examples of Nix expressions.
Each one is followed by the expected evaluation result.

The following example is a Nix expression adding two numbers:

```nix
1 + 2
```

    3

#### Interactive evaluation

Use [`nix repl`][nix-repl] to evaluate Nix expressions interactively (by typing them on the command line):

```console
nix repl
```

    Welcome to Nix 2.5.1. Type :? for help.

```console
nix-repl> 1 + 2
```

    3

:::{note}

> The Nix language by default uses lazy evaluation, and will only compute values when needed.

Author: I think this is a close-enough characterization for an introduction. Such details can be left for the reference manual.

#### Evaluating Nix files

Use [`nix-instantiate --eval`][nix-instantiate] to evaluate the expression in a Nix file.

```console
echo 1 + 2 > file.nix

nix-instantiate --eval file.nix
```

    3

<details><summary>Detailed explanation</summary>

The first command writes `1 + 2` to a file `file.nix` in the current directory.
The contents of `file.nix` are now `1 + 2`, which you can check with

```console
cat file.nix
```

    1 + 2

The second command runs `nix-instantiate` with the `--eval` option on `file.nix`, which reads the file and evaluates the contained Nix expression.
The resulting value is printed as output.

`--eval` is required to evaluate the file and do nothing else.
If `--eval` is omitted, `nix-instantiate` expects the expression in the given file to evaluate to a special value called *Derivation*, which we will come back to at the end of this guide in [](derivations).

If you do not need `file.nix` any more, remove it with

```console
rm file.nix
```

</details>

:::{note}
`nix-instantiate --eval` will evaluate `default.nix` if no file name is specified.

```console
echo 1 + 2 > default.nix

nix-instantiate --eval
```

    3
:::

:::{note}

> The Nix language by default uses lazy evaluation, and will only compute values when needed.

Author: Yes, intentional.
a) I expect people to scroll over this introductory boilerplate because they'd pick the `nix-repl` approach.
b) The invocation here is different and I think motivating it place is meaningful.

Nix language data types *without functions* work just like their counterparts in JSON and look very similar.
:::

<table>
<tr>
  <th>Nix language</th>
  <th>JSON</th>
</tr>
<tr>
<td>

```nix
{
  string = "hello";
  integer = 1;
  float = 3.141;
  bool = true;
  null = null;
  list = [ 1 "two" false ];
  attribute-set = {
    a = "hello";
    b = 2;
    c = 2.718;
    d = false;
  }; # comments are supported
}
```

</td>
<td>

```json
{
  "string": "hello",
  "integer": 1,
  "float": 3.141,
  "bool": true,
  "null": null,
  "list": [1, "two", false],
  "object": {
    "a": "hello",
    "b": 1,
    "c": 2.718,
    "d": false
  }
}
```

</td>
</tr>
</table>

:::{note}
- Attribute names usually do not need quotes.[^1]
- List elements are separated by white space.[^2]
:::

[^1]: Details: Nix manual - attribute naming rules <!-- TODO: create and link manual section -->
[^2]: Details: [Nix manual - lists][manual-lists]

[manual-lists]: https://nixos.org/manual/nix/stable/language/values.html#list

(rec-attrset)=
#### Recursive attribute set `rec { ... }`

You will sometimes see attribute sets declared with `rec` prepended.

> This allows access to attributes from within the set.

Author: I don't go into dynamic attributes because they are rare enough in practically relevant code not to matter here. If you just learn about Nix and want to go through some tutorials here you won't dig into Nixpkgs internals or NixOS module implementations where those become more prominent.

Example:

```nix
rec {
  one = 1;
  two = one + 1;
  three = two + 1;
}
```

    { one = 1; three = 3; two = 2; }

:::{note}

> Attribute sets are unordered.

Author: Out of curiosity, how is stable output relevant in Nix's domain?

In the following example we use the `let` expression to form a list:

```nix
let
  b = a + 1;
  c = a + b;
  a = 1;
in [ a b c ]
```

    [ 1 2 3 ]

</details>

Only expressions within the `let` expression itself can access the newly declared names.
We say: the bindings have local scope.

Counter-example:

```nix
{
  a = let x = 1; in x;
  b = x;
}
```

    error: undefined variable 'x'

           at «string»:3:7:

                2|   a = let x = 1; in x;
                3|   b = x;
                 |       ^
                4| }

<!-- TODO: exercise - use let to reuse a value in an attribute set -->

### Attribute access

Attributes in a set are accessed with a dot (`.`) and the attribute name.

Example:

```nix
let
  attrset = { x = 1; };
in
attrset.x
```

    1

Accessing nested attributes works the same way.

Example:

```nix
let
  attrset = { a = { b = { c = 1; }; }; };
in
attrset.a.b.c
```

    1

The dot (`.`) notation can also be used for assigning attributes.

Example:

```nix
let
  attrset = { a.b.c = 1; };
in
attrset
```

    { a = { b = { c = 1; }; }; }

(with)=

> ### `with ...; ...`

Author: We will not note this here, because it will not matter much for beginners.

### File system paths

The Nix language offers convenience syntax for file system paths.

Absolute paths always start with a slash (`/`).

Example:

```nix
/absolute/path
```

    /absolute/path

Paths are relative when they contain at least one slash (`/`) but to not start with one.
They evaluate to the path relative to the file containing the expression.

The following examples assume the containing Nix file is in `/current/directory` (or `nix repl` is run in `/current/directory`).

Example:

```nix
./relative
```

    /current/directory/relative

Example:

```nix
relative/path
```

    /current/directory/relative/path

One dot (`.`) denotes the current directory within the given path.

You will often see the following expression, which specifies a Nix file's directory.

Example:

```nix
./.
```

    /current/directory

<details><summary>Detailed explanation</summary>

Since relative paths must contain a slash (`/`) but must not start with one, and the dot (`.`) denotes no change of directory, the combination `./.` specifies the current directory as a relative path.

</details>

Two dots (`..`) denote the parent directory.

Example:

```nix
../.
```

    /current

> 

Author: This is true and so rarely seen that it's unimportant here.

Example:

```nix
let
  f = {a, b}: a + b;
in
f { a = 1; b = 2; }
```

    3

Counter-example:

```nix
let
  f = {a, b}: a + b;
in
f { a = 1; b = 2; c = 3; }
```

    error: 'f' at (string):2:7 called with unexpected argument 'c'

           at «string»:4:1:

                3| in
                4| f { a = 1; b = 2; c = 3; }
                 | ^
                5|

<!-- TODO: not the same as x: x.a + x.b (!!!!) -->

#### Default values

Also known as “default arguments”.

Destructured arguments can have default values for attributes.

This is denoted by separating the attribute name and its default value with a question mark (`?`).

Attributes in the argument are not required if they have a default value.

Example:

```nix
let
  f = {a, b ? 0}: a + b;
in
f { a = 1; }
```

    1

Example:

```nix
let
  f = {a ? 0, b ? 0}: a + b;
in
f { } # empty attribute set
```

    0

#### Additional attributes

Additional attributes are allowed with an ellipsis (`...`):

    {a, b, ...}: a + b

Unlike in the previous counter-example, passing an argument that contains additional attributes is not an error.

Example:

```nix
let
  f = {a, b, ...}: a + b;
in
f { a = 1; b = 2; c = 3; }
```

    3

> ### Named attribute set argument

Reviewer 5: Note that while formals definitions can refer to each other (e.g. `{ a, b ? a * 2 }: ...`) they can not refer to the named pattern (e.g. `{ a, b ? args.a }@args: ...` is invalid)

Author: Important only for writing, not reading code.

There is only one side effect in the Nix language that is relevant here:
reading files from the file system as *build inputs*

:::{note}
Side effects are performed while evaluating a given impure expression.
:::

Build inputs are files that build tasks refer to in their precise description of how to derive new files.

Since Nix language evaluation is otherwise pure, the only way to specify build inputs is explicitly with:

- File system paths
- Dedicated impure functions.

When run, a build task will only have access to explicitly declared build inputs.

:::{important}
Purity is the key to reproducible builds.

It precludes build tasks from referring to files which are not explicitly specified as build inputs.

Nix supports other types of impure expressions, such as [search paths](search-path) or the constant [`builtins.currentSystem`](https://nixos.org/manual/nix/stable/language/builtin-constants.html#builtins-currentSystem).
We do not cover those here more detail, as they do not matter for how the Nix language works in principle, and because they are discouraged for the very reason of breaking reproducibility.
:::

### Paths

> Whenever a file system path is rendered to a character string with [antiquotation](antiquotation), the contents of that file are copied to a special location in the file system, the *Nix store*, as a side effect.

Reviewer 5: The link syntax is broken here

Author: If you find bugs, please leave suggestions I can merge immediately. Otherwise I'll have to dig through each comment individually to figure out what exactly was wrong and how to fix it.

[^inherit-parens]: [`inherit`](inherit) is the only exception where parentheses are a syntactical requirement.

(names-values)=
## Names and values

Values in the Nix language can be primitive data types, lists, attribute sets, and functions.

We show primitive data types and lists as examples in the context of [attribute sets](attrset).
Later in this section we cover special features of character strings: [antiquotation](antiquotation), [file system paths](file-system-paths), and [indented strings](indented-strings).
We deal with [functions](functions) separately.

There are two ways to assign names to values in Nix: [attribute sets](attrset) and [`let` expressions](let).

Assignments are denoted by a single equal sign (`=`).

Whenever you see an equal sign (`=`) in Nix language code:
- On its left is the assigned name.
- On its right is the value, delimited by a semicolon (`;`).

(attrset)=
### Attribute set `{ ... }`

An attribute set is a collection of name-value-pairs, where names must be unique.

The following example shows all primitive data types, lists, and attribute sets.

:::{note}
If you are familiar with JSON, imagine the Nix language as *JSON with functions*.

> Nix language data types *without functions* work just like their counterparts in JSON and look very similar.

Reviewer 6: This is not true. It's fine to say *think about* Nix as JSON plus functions, but this is *only* an analogy. Specifically:

* Strings in C++ Nix are byte sequences and not Unicode-encoded in some shape or form as in JSON.
* Integers in Nix are 64 bit integers and not represented as floating point numbers as they are in JSON.
* Floats overall behave weirdly in Nix to the point that you should warn about them, i.e. they get truncated when serializing them in many situations.
* Due to the nature of Nix's strings, not every legal Nix attribute name is a legal JSON object key.

Reviewer 7: I think the formulation "imagine the Nix language as" is somehow similar to saying "think about Nix as", isn't it?

Reviewer 6: Yes, but the claim in the second sentence is stronger than that and wrong, as I've pointed out. “Nix language data types *without functions*” does *not* work just like their counterparts in JSON.

Reviewer 6: I feel that it is at least implied. With tutorials as the first contact point, one should be careful to avoid creating wrong impressions in readers – even by accident. FWIW my point is not that these differences *must* be explained, but rather that we should be careful to word it in a way that we don't imply they don't exist.

---

I'm also not sure these edges are that far out.

The integer thing is a relatively well known JSON pitfall – Nix is not thorny here as it behaves like you'd expect.

```console
> echo 34343499993845656749 | jq
34343499993845658000
```

The float situation can be discovered quite easily. E.g.:

```console
nix-repl> 1.000004 
1
nix-repl> builtins.toJSON 1.000004
"1"
```

String indexing, I'd say is a very basic topic that should be covered in a tutorial, since the internal representation is apparent in this, it is relevant. Indexing into non-ASCII strings will unexpectedly  e.g. `builtins.substring 0 1 "καλημερα"` will return in some garbage being returned. E.g. “Nix strings are similar to C strings” is a statement like the one about JSON data, but more accurate.

    { a = 1; b = 2; }

## Functions

The Nix language is Turing-complete.
It gives you the power to tame complexity and the freedom to create overwhelming amounts of it.

Therefore, it is recommended to use functions sparingly and instead think more in terms of data, because data is usually easier to understand.

### Arguments

Nix functions take exactly one argument.

    x: x + 1

Argument and function body are separated by a colon `:`.

Applying a function to an argument means writing the argument after the function name.

Example:

    nix-repl> let
              f = x: x + 1;
              in
              f 1
    2

Arguments can be chained.

    x: y: x + y

The above function takes one argument and returns a function `y: x + y` with `x` set to passed value. 

Example:

    nix-repl> let
              f = x: y: x + y;
              in
              f 1
    «lambda @ (string):2:8»

The `lambda` indicates the resulting value is a function.

Applying that to another argument yields the inner body `x + y`, which can now be fully evaluated.

    nix-repl> let
              f = x: y: x + y;
              in
              f 1 2
    3

### Keyword arguments

> 

Author: ```suggestion

Also known as “attribute set argument”.
```

@infinisil This sounds more correct, but I haven't seen that one yet. Is it common? I would just pose it as the main way to call it, but it sounds clunky to me.

Nix functions can explicitly take an attribute set as argument.

    {a, b}: a + b

> This is equivalent to
> 
>     x: x.a + a.b
> 
> The argument defines the exact attributes that have to be in that set.
> Leaving out or passing additional attributes is an error.

Author: ```suggestion
While this is similar to

    x: x.a + x.b

only keyword arguments define the exact attributes that have to be in that set.
Leaving out or passing additional attributes is an error.
```

Bu we may as well leave out the similarity, as it introduces more detail than strictly necessary. I thought it may help readers transition from what may be structure of the argument in a function declaration.

Example:

    nix-repl> let
              f = {a, b}: a + b
              in
              f { a = 1; b = 2; }
    3

### Default attributes

Also known as “default arguments”.

Arguments can have default values for attributes, denoted with a question mark “?”.

    {a, b ? 0}: a + b

Attributes in the argument are not required if they have a default value.

Example:

    nix-repl> let
              f = {a, b ? 0}: a + b
              in
              f { a = 1; }
    1

Example:

    nix-repl> let
              f = {a ? 0, b ? 0}: a + b
              in
              f { } # empty attribute set
    0

### Additional attributes

You can allow additional attributes with an ellipsis (`...`):

    {a, b, ...}: a + b

Example:

    nix-repl> let
              f = {a, b, ...}: a + b
              in
              f { a = 1; b = 2; c = 3; }
    3

### Named keyword arguments

Also known as “@ syntax” or “‘at’ syntax”.

>     {a, b, ...}@args: a + b + args.x
> 

Author: ```suggestion
    {a, b, ...}@args: a + b + args.x

or

    args@{a, b, ...}: a + b + args.x
```

# Reading the Nix language

The Nix language is used to declare packages and configurations for the Nix package manager.

You will quickly encounter Nix language expressions that may look very complicated.
Yet, the language has only few basic concepts that can be combined arbitrarily.
Use this guide to learn reading code in Nix language and understand its structure.

> Imagine the Nix language as JSON with functions.

Author: This is a deliberate oversimplification to give . @edolstra just recently said [the very same thing](https://youtu.be/h8hWX_aGGDc?t=955) and I only wrote "functions" not to scare off people who are not familiar with lambda calculus.

To what extent is it misleading? What would you suggest instead as a bird's-eye-view on the topic?

Its purpose is to define structured data, and functions help with conveniently producing more complex data.
All of the language constructs can be composed arbitrarily.

### Attribute sets

> The primary data structure is the attribute set, a collection of key-value pairs.

Author: > Calling it the primary data structure feels weird

I wanted to say it's the most important one, because everything revolves around attrsets. See [JSON+lambdas](https://github.com/NixOS/nix.dev/pull/267/files#r932169165). 

> Also data structure is a bit off

Could you clarify why, just so I understand?

(In general I agree that using the most precise terms is more correct. Yet, I don't think its also more better, as the goal of these tutorials is to pick up developers, who are not necessarily all too familiar with PLT, from where they are). 

> better distinction between values and expressions. This document here mixes these two concepts fairly arbitrarily.

Could you please provide concrete suggestions?

See attribute [naming rules]().

Nix has additional primitives:

### String

#### Multiline string

    nix-repl> ''
              string
              spanning
              multiple
              lines
              ''
    "string\nspanning\nmultiple\nlines\n"

Whitespace that prefixes all lines is stripped automatically.

Multiline strings ease the use of quotes and special characters.

    nix-repl> ''
              "double quoted" and 'single quoted'
              ''
    "\"double quoted\" and 'single quoted'\n"

#### String interpolation

    nix-repl> let
              name = "Nix";
              in
              "hello ${name}"
    "hello Nix"

See escaping rules.

### Path

#### Relative path

    ./relative/path

The path is relative to the file containing the expression.

#### Absolute path

    /absolute/path

#### Home directory path

    ~/path/in/home/directory

#### Named paths

Also known as “angle bracket syntax”.

    <name>

Examples:

    # write a Nix expression to a file
    echo 123 > nix_path_file

    # set `NIX_PATH` to point a name to that file
    env NIX_PATH=name_in_nix_path=nix_path_file nix repl

>     nix repl> import <nix_path_file>

Reviewer 8: `import <name_in_nix_path>`?

Author: ```suggestion
    nix repl> import <name_in_nix_path>
```

    123

Nix simply reads and parses the `NIX_PATH` environment variable.
Entries are separated by colons (`:`).

> Note: The value of a named path depends on external system state. Therefore it is recommended not to use them.

Reviewer 8: As a new nix user, I find comments like this a bit scary because I see examples (especially ones involving `<nixpkgs>`) all over the place. Are these examples doing it wrong? Or should the "recommended not to use them" advice be qualified?

Author: > Are these examples doing it wrong?

IMHO: Yes. At best it's a sloppy short-hand. See the [pinning tutorial](https://github.com/NixOS/nix.dev/blob/21ed44040211f92edc179f1017ab96a74a4ebdd7/source/tutorials/towards-reproducibility-pinning-nixpkgs.md?plain=1#L5-L18) for reasoning.

```suggestion
Note:
The value of a named path depends on external system state.
That means, the same Nix expression can produce different results, which may be very confusing.
Therefore {ref}`we recommend to avoid named paths<Towards reproducibility: pinning Nixpkgs>`.
```

#### Path concatenation

    /nix + /store

Paths can be concatenated with each other or with strings.

Examples:

    nix-repl> /nix + /store
    /nix/store

    nix-repl> /nix/var + "/nix"
    /nix/var/nix

### Let bindings

    nix-repl> let a = 1; in a + a
    2

Let bindings allow declaring named values for repeated use.

### Recursive sets

Recursive sets allow elements to refer to attributes within the set.
They are constructed with the `rec` keyword.

    nix-repl> rec { a = 1; b = a + a; }
    { a = 1; b = 2; }

Note: This construct
- allows infinite recursion, such as `rec { x = x; }.x`
- can quickly become hard to understand

> - does not lend itself to static analysis.

Author: ```suggestion
```

> - can quickly become hard to understand
> - does not lend itself to static analysis.

Author: ```suggestion
- can quickly become hard to understand.
```

Therefore it is recommended to use let bindings instead.

    nix-repl> let x = 1; in { a = x; b = x + x; }
    { a = 1; b = 2; }

## Functions

The Nix language is Turing-complete.
It gives you the power to tame complexity and the freedom to create overwhelming amounts of it.

> Therefore, it is recommended to use functions sparingly and instead think more in terms of data, because data is usually easier to understand.

Author: Okay, this is really opinionated, and we may want to discuss this on the docs team meeting, but here's my stance:

Independently, @edolstra and @ndmitchell aptly referred to what people tend create in Turing-complete configuration languages as "monstrosities". In my opinion, that is what people tend to create in *any* programming language if they don't care enough about long-term maintenance (which often has good reasons) and still scale up the code base (which is the curse of success). Both Eelco and Neil opted for restricting the language for Buck v2 and Nix respectively, by introducing a data-oriented layer at the very top. I argue we could instead encourage people to write code more carefully from the start. That is the entire premise of the article - JSON with functions where it helps.

You can allow additional attributes with an ellipsis (`...`):

    {a, b, ...}: a + b

Example:

    nix-repl> let
              f = {a, b, ...}: a + b
              in
              f { a = 1; b = 2; c = 3; }
    3

### Named keyword arguments

Also known as “@ syntax” or “‘at’ syntax”.

    {a, b, ...}@args: a + b + args.x

where additional attributes are subsumed under a name.

Example:

    nix-repl> let
              f = {a, b, ...}@args: a + b + args.x
              in
              f { a = 1; b = 2; c = 3; x = 5 }
    8

This can be useful if this remaining attribute set needs to be processed as a whole.

[^1]: For example, the built-in function `throw` causes evaluation to stop. However, the following works fine, because `throw` is never evaluated:
     nix-repl> let lazy = { a = "success"; b = builtins.throw "error"; }; in lazy.a
"success"

## Summary

As a programming language, Nix is

- *declarative*

  It has no notion of executing sequential steps.
  Dependencies between operations are established only through data.
  Everything in Nix is an expression that results in a single value.
  Every Nix file (`.nix`) contains a single expression.

- *purely functional*

>   Pure means: Nix does not change the value of declarations during computation – there are no variables, only names for immutable values.

Reviewer 2: > there are no variables

Sounds very weird, because there are, but they aren't variable. Maybe reword to "There are no mutable variables"

Author: "variables that are not variable" is an oxymoron. Therefore there are no variables in the Nix language.

Reviewer 2: Even if variables never change during execution, they are still called variables in the sense that their value can vary between executions. See https://en.wikipedia.org/wiki/Variable_(mathematics) or https://www.merriam-webster.com/dictionary/variable

Reviewer 4: I think a focus on whether Nix does or does not have "actual variables" is a confusing distraction when describing what pure functional programming means.

Colloquial use of the word "variable", at least among English speakers, applies in an intuitive way to functional programming. Nix has these in function arguments and let bindings, both of which allow a program to cause one name to take on different values. This is the core concept behind what a "variable" is.

See also https://nixos.org/manual/nix/stable/expressions/language-constructs.html#let-expressions which says "A let-expression allows you to define local variables for an expression."

See also https://nixos.org/guides/nix-pills/basics-of-language.html#idm140737320512656 which shows you how to use a let expression to shadow variables, effectively making one name refer to multiple values by introducing a new scope.

There is also https://en.wikipedia.org/wiki/Variable_(computer_science) which says "In pure functional programming, variables are bound to expressions and keep a single value during their entire lifetime due to the requirements of referential transparency."

I would suggest that the description of "pure" avoid the problem entirely by not focusing on a "variables that are not variable" oxymoron. Instead, focus on the idea that functions always return the same when passed the same arguments, and that computed values cannot be modified.  Or something like that...

Author: I will remove this section entirely. I had been influenced by the manual, [which says](https://web.archive.org/web/20220716000213/https://nixos.org/manual/nix/stable/expressions/expression-language.html)

> for instance, there is no variable assignment

The manual should be the reference on what the Nix language is and does, so I moved my attempt to describe it more clearly there: https://github.com/NixOS/nix/pull/6907

The [NixOS Wiki on Nix language replicates this](https://web.archive.org/web/20220812152834/https://nixos.wiki/wiki/Overview_of_the_Nix_Language), and that redundancy should IMHO go away, and instead we should always refer to the manual. Same for this tutorial.

''
```

    "one\n two\n  three\n"

See [escaping rules]().

# Using existing functions

<!-- TODO: built-ins and library -->

# Building software using side effects

So far we have only covered what we call *pure expressions*:
declaring data and transforming it with functions.

Building software requires interaction with the outside world, which we call *side effects*.

There are two main side effects in the Nix language that are relevant here:
1. Reading files from the file system as build inputs
2. Writing files to the file system as build outputs

<!-- TODO: side effects - fetchers and derivations -->

See Nix Pills for a detailed explanation on how Nix the package manager builds software using library functions and packages from the Nix package collection.

# Summary

You should now be able to read Nix language code for simple packages and configurations, and come up with similiar explanations of the following examples.

Example:

```nix
{ pkgs ? import <nixpkgs> {}, ... }:
pkgs.mkShell {
  # ...
}
```

Explanation:

This expression is a function that takes an attribute set as an argument.
If the argument has the attribute `pkgs`, it will be used in the function body.

> Otherwise, the default value of importing from the search path `<nixpkgs>` and calling the resulting function with an empty attribute set will be used.

Reviewer 9: ⚠️ You didn't explain `import` yet, what it does, and how to use it

Author: There will be a section on import, this PR is still a draft.
